### DISCUSSION TITLE
StructuredArguments to return Marker instead of interface

### DISCUSSION THREAD

**Discussion by Jereminet** [logfellow/logstash-logback-encoder#524] [Q&A] [answered]
Hi,

When using both slf4j logger and StructuredArguments, I realised that if I want to log a message and enhance the JSON event with the fields from my POJO like this
```
var person = new Person("John", "Doe");
logger.info("New person created", fields(person));
```
I get a warning telling me that I didn't provide any placeholder in the message for the fields, but I don't want this. Indeed, it would be redundant to have the fields added both in the message and as separated fields in my JSON.
Therefore, I wanted to leverage the support for Marker and instead call my logger like this
```
logger.info(fields(person), "New person created");
```
which gives me a compilation error because it excepts a Marker. Thing is that this is indeed what is returned by your function but the specification of the method is signed with the interface, not the Marker type. It therefore requires a cast at user level which is, as you could imagine, not really clean.

Is there a specific reason for this design ? Do you think it could be possible to return the concrete type instead of the interface ?

As an example, a possible fix for the _fields()_ method would be to switch from
```
public static StructuredArgument fields(Object object) {
    return new ObjectFieldsAppendingMarker(object);
}
```
to
```
public static ObjectFieldsAppendingMarker fields(Object object) {
    return new ObjectFieldsAppendingMarker(object);
}
```

Regards,

Jeremy

**Comment by philsttr** [comment, marked as answer]
Hi Jeremy,

When using markers, use the static methods in `net.logstash.logback.marker.Markers` instead of the ones in `net.logstash.logback.argument.StructuredArguments`

**Reply by Jereminet** [reply]
So used to play with StructuredArguments that I forgot to look at your beautiful documentation !
Thanks for you response, I'll surely use Markers instead.

Have a nice day :)
